h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -171,6 +173,11 @@ public boolean stopEarly(Model model, ScoringInfo[] sk) {
      */
     final protected long _maxHyperSpaceSize;
 
+    /**
+     * Provides filtering logic to make walking process more efficient.
+     */
+    PermutationFilter<MP> _permutationFilter;
```

Comment:
this field is declared on BaseWalker but set in each subclass, thats quite smelly
I would suggest two possible improvements
a) make this field private, implement iterator() method in base walker and make it final move the if (filter != null) check here and create new abstract method makeIterator() and rename the current implementations of iterator() to makeIterator()
or
b) move all this wrap-iterator-in-filter logic out of BaseWalker into a separate class FilteringWalker that would wrap around any HyperSpaceWalker implementation (I prefer this one since it better separation of concerns)

Reply:
b) +1

Reply:
I just wonder what will be the next improvement for separation of concerns: FilteringGridSearch class, h2o-core-filtering module, h2o-3 fork for filtering? :) Let's see, will try to do b)

Reply:
@honzasterba done. Please review the move of the logic into FilteredWalker. I will check one more time for javadocs and other things that might have become inconsistent due to the move.